program/src/guards/bot_tax.rs

```diff
@@ -39,53 +40,17 @@ impl Condition for BotTax {
-            let instruction_sysvar_account_info = instruction_sysvar_account.to_account_info();
-            let instruction_sysvar = instruction_sysvar_account_info.data.borrow();
-            // the next instruction after the mint
-            let next_ix = get_instruction_relative(1, &instruction_sysvar_account_info);
+            let ix_sysvar_account = &ctx.accounts.instruction_sysvar_account;
+            let ix_sysvar_account_info = ix_sysvar_account.to_account_info();
 
-            if let Ok(ix) = next_ix {
-                let discriminator = &ix.data[0..8];
-                let after_collection_ix =
-                    get_instruction_relative(2, &instruction_sysvar_account_info);
-
-                if !cmp_pubkeys(&ix.program_id, &crate::id())
-                    || discriminator != [103, 17, 200, 25, 118, 95, 125, 61]
-                    || after_collection_ix.is_ok()
-                {
-                    // we fail here, it is much cheaper to fail here than to allow a malicious user
-                    // to add an ix at the end and then fail
-                    msg!("Failing and halting due to an extra unauthorized instruction");
-                    return err!(CandyGuardError::MintNotLastTransaction);
-                }
+            // the next instruction after the mint
+            if get_instruction_relative(1, &ix_sysvar_account_info).is_ok() {
+                msg!("Failing and halting due to an extra unauthorized instruction");
+                return err!(CandyGuardError::MintNotLastTransaction);
             }
 
-            let mut idx = 0;
-            let num_instructions = read_u16(&mut idx, &instruction_sysvar)
-                .map_err(|_| ProgramError::InvalidAccountData)?;
-
-            for index in 0..num_instructions {
-                let mut current = 2 + (index * 2) as usize;
-                let start = read_u16(&mut current, &instruction_sysvar).unwrap();
-
-                current = start as usize;
-                let num_accounts = read_u16(&mut current, &instruction_sysvar).unwrap();
-                current += (num_accounts as usize) * (1 + 32);
-                let program_id = read_pubkey(&mut current, &instruction_sysvar).unwrap();
-
-                if !cmp_pubkeys(&program_id, &crate::id())
-                    && !cmp_pubkeys(&program_id, &::spl_token::id())
-                    && !cmp_pubkeys(
-                        &program_id,
-                        &anchor_lang::solana_program::system_program::ID,
-                    )
-                    && !cmp_pubkeys(&program_id, &spl_associated_token_account::ID)
-                {
-                    msg!("Transaction had ix with program id {}", program_id);
-                    return err!(CandyGuardError::MintNotLastTransaction);
-                }
-            }
+            // verifies that only authorized programs have instructions in the transaction
+            verify_programs(ix_sysvar_account_info, DEFAULT_PROGRAMS)?;
```

Comment:
sorta. Makes a bit more sense than all the code duplication though. For now it has to duplicate the same behavior to prevent a breaking change. If we want to remove it we'd have to make an announcement I think?

Reply:
It is just to avoid code duplication. I have a plan to remove it from the `bot tax`. There is now a `verify` function on each guard that allows the guard to validate its settings. We can prevent setting the `last_instruction` to `true` on the `bot tax` for new candy guard accounts, so eventually all accounts will need to use the new guard (and a new `LastInstruction` guard).